Sources/GRPC/Docs.docc/Proposals/0001-stub-api.md

```diff
@@ -0,0 +1,1100 @@
+
+  /// Metadata associated with the error.
+  public var metadata: Metadata
+
+  public init(code: Code, message: String, metadata: Metadata = [:]) {
+    // ...
+  }
+}
+```
+
+### Generated server code
+
+Code generated for each service includes two protocols. The first, higher-level
+protocol which most users interact with, includes one function per defined
+method. The shape of the function matches the method definition. For example
+unary methods accept a `ServerRequest.Single` and return a
+`ServerResponse.Single`, bidirectional streaming methods accept a
+`ServerRequest.Stream` and return a `ServerResponse.Stream`.
+
+The second, base protocol, defines each method in terms of streaming requests
+and responses. The higher-level protocol refines the base protocol and provides
+default implementations of methods in the base protocol in terms of their
+higher-level counterpart.
+
+The base protocol is an escape hatch allowing advanced users to have further
+control over their RPCs. As an example, if a service owner needs to respond to
+initial metadata in a client streaming RPC before processing the complete stream
+of messages from the request they could implement their RPC in terms of the
+fully streaming version provided by the base protocol.
+
+Users can throw any error from each method. Since gRPC has a well defined error
+model, gRPC Swift catches errors of type `RPCError` and extracts the code and
+message. The code and message are propagated back to the client as the status of
+the RPC. The library discards all other errors and returns a status with code
+`unknown` to the client.
+
+The following code demonstrates how these protocols would look for the Echo
+service. Some details are elided as they aren't relevant.
+
+```swift
+// (Defined elsewhere.)
+public typealias EchoRequest = ...
+public typealias EchoResponse = ...
+
+/// The generated base protocol for the "Echo" service providing each method
+/// in a fully streamed form.
+///
+/// This protocol should typically not be implemented, instead you should
+/// implement ``EchoServiceProtocol`` which refines this protocol. However, if
+/// you require more granular control over your RPCs then they may implement
+/// this protocol, or methods from this protocol, instead.
+public protocol EchoServiceStreamingProtocol: BindableService, Sendable {
+  func get(
+    request: ServerRequest.Stream<EchoRequest>
+  ) async throws -> ServerResponse.Stream<EchoResponse>
+
+  func collect(
+    request: ServerRequest.Stream<EchoResponse>
+  ) async throws -> ServerResponse.Stream<EchoResponse>
+
+  func expand(
+    request: ServerRequest.Stream<EchoRequest>
+  ) async throws -> ServerResponse.Stream<EchoResponse>
+
+  func update(
+    request: ServerRequest.Stream<EchoResponse>
+  ) async throws -> ServerResponse.Stream<EchoResponse>
+}
+
+// Generated conformance to `BindableService`.
+extension EchoServiceStreamingProtocol {
+  public func bind(to router: inout RPCRouter) {
```

Comment:
Is `bind` a term of are here in gRPC? I personally think `register(with router:)` might sound better.

Reply:
Not especially, it fits with `BindableService` though. The naming isn't really important -- users are very unlikely to call this directly. We'd probably have API on the server/router to add a service which would just call through to this.